### PR TITLE
WIP - CLIMS-421 - transition model

### PR DESCRIPTION
**Pull request by mariya**
This is a WIP because I have some questions that need answering before this is merged. @withrocks please see my comments, feedback is appreciated.

### CODE REVIEW

File: src/clims/models/transition.py [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+    SPAWN = 1
+
+
+class Transition(Model):
+    """
+    Represents a Transition for a Substance.
+    """
+    __core__ = True
+
+    BACKEND_CAMUNDA = 'camunda'
```
Comment (mariya) [inline src/clims/models/transition.py]: What does this do?

File: src/clims/models/transition.py [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+
+    # Added because it seems to be used in all CLIMS models.
+    # TODO: check if it is actually needed?
+    organization = FlexibleForeignKey('sentry.Organization')
```
Comment (mariya) [inline src/clims/models/transition.py]: Will all our CLIMS models link to an Organization object? What is the purpose?

File: src/clims/models/transition.py [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+
+    # Added in case we want to be able to audit a user's history directly.
+    user = FlexibleForeignKey('sentry.User', related_name='transitions')
+
+    # The work batch in which this transition was created, if it was created in a work batch context.
+    # TODO: should we disallow nulls here? Are transitions ever created outside a work batch context?
+    work_batch = FlexibleForeignKey('WorkBatch', null=True, related_name='transitions')
```
Comment (mariya) [inline src/clims/models/transition.py:22]: Per comment: should we disallow nulls here? Are transitions ever created outside a work batch context?

File: src/clims/models/transition.py [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+    source_substance = FlexibleForeignKey('Substance', related_name='transitions')
+    target_substance = FlexibleForeignKey('Substance', related_name='transitions')
+
+    transition_type = models.IntegerField(default=TransitionType.MOVE)
+
+    class Meta:
+        app_label = 'clims'
+        db_table = 'clims_transition'
+
+    __repr__ = sane_repr('work_batch_id', 'source_location_id', 'source_substance_id', 'target_location_id', 'target_substance_id')
```
Comment (mariya) [inline src/clims/models/transition.py:33]: How is sane_repr used? It seems that this is a sentry wrapper function for representing an object as a string, but is this representation actually used somewhere?

Comment (mariya) [reply src/clims/models/transition.py:33]: Spoke with Edvard now. We're guessing this is used for logging and debugging?

File: src/clims/models/transition.py [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+    organization = FlexibleForeignKey('sentry.Organization')
+
+    # Added in case we want to be able to audit a user's history directly.
+    user = FlexibleForeignKey('sentry.User', related_name='transitions')
+
+    # The work batch in which this transition was created, if it was created in a work batch context.
+    # TODO: should we disallow nulls here? Are transitions ever created outside a work batch context?
+    work_batch = FlexibleForeignKey('WorkBatch', null=True, related_name='transitions')
+
+    source_location = FlexibleForeignKey('SubstanceLocation', related_name='source_transitions')
```
Comment (mariya) [inline src/clims/models/transition.py:24]: Are reverse accessors here a good idea, or should they be removed?

File: src/clims/models/transition.py (hunk 1 above)
Comment (withrocks) [inline src/clims/models/transition.py]: This constant should probably not be here?

File: src/clims/models/transition.py (hunk 2 above)
Comment (withrocks) [inline src/clims/models/transition.py]: I think that this doesn't need to be in the model, although a transition is under an organization in the end. The reason is that the transitions always happen in the scope of a `WorkBatch` and they relate to the organization.

File: src/clims/models/transition.py [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+    target_location = FlexibleForeignKey('SubstanceLocation', related_name='target_transitions')
+    source_substance = FlexibleForeignKey('Substance', related_name='source_transitions')
```
Comment (withrocks) [inline src/clims/models/transition.py]: I think only the source_location and target_location are really needed here, as they include a link to the substance (and version) in it.

So you can think of `SubstanceLocation` as a pointer to a substance, it knows where it resides.

**Comment by withrocks** [comment]
Looks cool! I have a few comments.

File: src/clims/models/transition.py (hunk 2 above)
Comment (mariya) [reply src/clims/models/transition.py]: Removing

File: src/clims/models/transition.py (hunk 6 above)
Comment (mariya) [reply src/clims/models/transition.py]: That's useful feedback, thanks. Removing.

**Pull request merged**